src/adapters/marketplace-api-fetcher.ts

```diff
@@ -0,0 +1,438 @@
+      })
+      throw error
+    }
+  }
+
+  async function fetchUserEmotes(
+    address: string,
+    params?: MarketplaceApiParams
+  ): Promise<{ emotes: OnChainEmote[]; total?: number }> {
+    try {
+      const baseEndpoint = `/v1/users/${address.toLowerCase()}/emotes/grouped`
+      const endpoint = buildEndpointWithParams(baseEndpoint, params)
+
+      let groupedEmotes: MarketplaceApiGroupedEmote[]
+      let total: number | undefined
+
+      if (params && (params.limit !== undefined || params.offset !== undefined)) {
+        // Use direct API call with parameters
+        const response = await makeApiRequest<MarketplaceApiResponse<MarketplaceApiGroupedEmote>>(endpoint)
+        // Validate response structure
+        if (!response || !response.data || !Array.isArray(response.data.elements)) {
+          throw new MarketplaceApiError(
+            `Invalid API response structure: expected response.data.elements to be an array`
+          )
+        }
+        groupedEmotes = response.data.elements
+        total = response.data.total
+      } else {
+        // Fallback to fetching all pages when no pagination specified
+        groupedEmotes = await fetchAllPages<MarketplaceApiGroupedEmote>(baseEndpoint)
+        total = groupedEmotes.length // For fallback, total is the actual count
+      }
+
+      return {
+        emotes: groupedEmotes.map(mapGroupedEmoteToOnChainEmote),
+        total
+      }
+    } catch (error) {
+      logger.error(`Failed to fetch user emotes for ${address}`, {
+        error: error instanceof Error ? error.message : String(error)
+      })
+      throw error
+    }
+  }
+
+  async function fetchUserNames(
+    address: string,
+    params?: MarketplaceApiParams
+  ): Promise<{ names: Name[]; total?: number }> {
+    try {
+      const baseEndpoint = `/v1/users/${address.toLowerCase()}/names`
+      const endpoint = buildEndpointWithParams(baseEndpoint, params)
+
+      let apiNames: MarketplaceApiName[]
+      let total: number | undefined
+
+      if (params && (params.limit !== undefined || params.offset !== undefined)) {
+        // Use direct API call with parameters
+        const response = await makeApiRequest<MarketplaceApiResponse<MarketplaceApiName>>(endpoint)
+        // Validate response structure
+        if (!response || !response.data || !Array.isArray(response.data.elements)) {
+          throw new MarketplaceApiError(
+            `Invalid API response structure: expected response.data.elements to be an array`
+          )
+        }
+        apiNames = response.data.elements
+        total = response.data.total
+      } else {
+        // Fallback to fetching all pages when no pagination specified
+        apiNames = await fetchAllPages<MarketplaceApiName>(baseEndpoint)
+        total = apiNames.length // For fallback, total is the actual count
+      }
+
+      return {
+        names: apiNames.map(mapApiNameToName),
+        total
+      }
```

Comment:
Do you think we can encapsulate the logic for checking if pagination is specified, or defaulting to fetching all elements when pagination is not provided, into another function? This way, the three exposed functions will be even cleaner.